rand_core/src/lib.rs

```diff
@@ -247,55 +247,138 @@ pub enum ErrorKind {
     __Nonexhaustive,
 }
 
-#[cfg(feature="std")]
+impl ErrorKind {
+    pub fn description(&self) -> &'static str {
+        match *self {
+            ErrorKind::Unavailable => "permanent failure or unavailable",
+            ErrorKind::Transient => "transient failure",
+            ErrorKind::NotReady => "not ready yet",
+            ErrorKind::Other => "uncategorised error",
+            ErrorKind::__Nonexhaustive => unreachable!(),
+        }
+    }
+}
+
 #[derive(Debug)]
-pub struct Error {
-    pub kind: ErrorKind,
-    pub cause: Option<Box<std::error::Error>>,
+enum ErrorDetails {
+    None,
+    Str(&'static str),
+    #[cfg(feature="std")]
+    Error(Box<std::error::Error + Send + Sync>),
 }
 
-#[cfg(not(feature="std"))]
-#[derive(Debug, Clone, PartialEq, Eq)]
+impl fmt::Display for ErrorDetails {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ::std::error::Error;
+        write!(f, "{}", self.description())
+    }
+}
+
+// Impl is only used for ErrorDetails::Str case, but can't implement for an enum case
+#[cfg(feature="std")]
+impl ::std::error::Error for ErrorDetails {
+    fn description(&self) -> &str {
+        match *self {
+            ErrorDetails::None => "NA", // case shouldn't be used
```

Comment:
It's better to write `unreachable!()` here if this case indeed can not be used anywhere.